### ISSUE TITLE
Rich Text Editor in avalonia

### ISSUE THREAD

**Issue by fengyinyong** [AvaloniaUI/Avalonia#4625]
I need a rich text editor in my avalonia project，just like the rich text editor in WPF, witch can add words、images.....
I have read the Issue https://github.com/AvaloniaUI/Avalonia/issues/4410 and  https://github.com/AvaloniaUI/AvaloniaEdit/.
But they do not fulfill my needs

**Comment by kekekeks** [comment]
FlowDocument text model is being worked on

**Comment by fengyinyong** [comment]
When work it out，please tell me in this issue. Thank you very much.

**Comment by dragonman117** [comment]
> FlowDocument text model is being worked on

Is there any chance we can get an indication of where this work is going on? I don't really see it on any of the roadmaps I've been able to find, and I don't see any branches that appear to be showing any active work for this feature.

**Comment by danwalmsley** [comment]
@Gillibald is actively working on it

**Comment by Gillibald** [comment]
https://github.com/AvaloniaUI/Avalonia/pull/5461

**Comment by warappa** [comment]
Is there any news on this topic as PR #5461 was abandoned?

And as demand was mentioned in the PR: How should one show demand? Via thumbs-up on this ticket? Somewhere in discussion section?

Anyway, thanks for your efforts!

**Comment by maxkatz6** [comment]
Rich TextBlock with inlines from #5461 was already added after multiple other PRs. As well as selectable TextBlock.
RichTextBox specifically wasn't yet done. Current text rendering with IME support is in higher priority. After that work on RichTextBox can be started. @Gillibald might know better estimates.
So, since it is already planned to be added earlier or sooner, I don't think there is a point in voting.

Meanwhile, AvaloniaEdit is still a working alternative for some use cases.

**Comment by JesseDietrichson** [comment]
@Gillibald @maxkatz6 Is there any way we could discuss bumping the priority of RichTextBox + Flowdocument. Root is interested in funding this effort.

Would love to chat about this more. My email is jdietrichson@rootapp.com

**Comment by Gillibald** [comment]
Please contact us at [team@avaloniaui.net](mailto:team@avaloniaui.net) and we can setup a meeting to discuss your requirements and what we can do to achieve them.

**Comment by parentelement** [comment]
Not sure if people are still waiting on this but I finally decided to just build my own.  I explored MAUI, which doesn't have one and doesn't have native Linux support, Uno doesn't have one, and Avalonia doesn't have one, so I started building my own based on my modified version of the RichTextKit library.  I'm still working on features and working through some bugs and I still to add some hooks and scroll bars but currently I support:
* Keybaord navigation and mouse interaction.
* Text Styles (Bold, Italic, Underline, Subscript, Superscript, etc... everything RichTextKit does)
* Merging styles (Applying a style to a selection merges that style with all runs in the selection - RichTextKit doesn't do this, it's a modification).
* Mouse wheel scrolling - code based scrolling.
* Copy/Paste/Undo/Redo
* Fonts, Font Sizes
* X and Y Scaling (zoom in/out)

There's a lot of work still to be done and I'm building on top of my own abstraction that'll allow me to build the same control for other platforms as well, but I'm starting with Avalonia.  Here's a brief rough video of it working.

[https://youtu.be/AKlW6Cl5SuU](url)

**Comment by timunie** [comment]
@parentelement that looks really awsome. If the control will be open source, consider to add it to awsome avalonia 👍

**Comment by parentelement** [comment]
@Gillibald 100%.  If I was developing purely for Avalonia I'd probably just go the route of porting the WPF controls directly, but I want to support Maui and Uno as well.  Fortunately, RichTextKit already essentially provides what I need because I'm only actually rendering what's within the visible bounds so rendering isn't an issue.  To add, though, Skia by itself isn't always fantastic for text.  RTK uses Harfbuzz for text shaping and ligature support which makes supporting the whole set of Unicode (and even emojis) possible more richly than what TextFormatter can provide.

**Comment by Gillibald** [comment]
TextFormatter supports Unicode 13 and is optimized for memory consumption. Its API is designed to be used to support arbitrary sources of text and supports random access so virtualization is no issue. I just wanted to inform you.

**Comment by parentelement** [comment]
Good to know.  I'll have to see if it's been ported for Maui and Uno as
well.

**Comment by parentelement** [comment]
Looks like TextFormatter is only ported to Avalonia, but I'm pretty happy with RTK.  Making some pretty decent progress.

https://youtu.be/Mutb0SWRjwQ

**Comment by doublecount** [comment]
@parentelement 
saw your great work on RichTextBox. Is there already any plan for release, what kind of license do you consider?

**Comment by parentelement** [comment]
> @parentelement saw your great work on RichTextBox. Is there already any plan for release, what kind of license do you consider?

It's in progress.  My focus has been cleaning up some bugs.  Tentatively, the plan is to release it as two separate packages.  One will be open source for sure. That have the RTK fork along with the core interface library and I'll have adapters for each target platform starting with Avalonia.

**Comment by hematec** [comment]
@parentelement 
you've done an impressive job on this!
Are there any news for a planned release date?

**Comment by parentelement** [comment]
> @parentelement you've done an impressive job on this! Are there any news for a planned release date?

Thanks, appreciate the kind words!  There's no specific time frame yet.  I'm still working on it but it's somewhat sporadic as I'm the only contributor currently.

**Comment by mateli** [comment]
@parentelement I suggest that you decide what to open source asap so that others can contribute. Also I do not think that a fully functioning rich text editor is required for making a first release. Instead deciding on a minimal viable functionality and making sure that future updates do not regress the controls behavior is useful.

On another note it may be useful to considering other rendering engines than Skia. [Vello](https://github.com/linebender/vello) for example seems really fast and it uses wgpu for rendering so it should work on even more platforms than Skia. Using wgpu for headless rendering and then putting the result into the Skia pipeline could be fast if it can be done within the GPU.

Although most of that would be premature. The first release should be something that can display and edit simple rich text.

**Comment by alexandrehtrb** [comment]
To be honest, I just need a TextBox that can show coloured parts of text

**Comment by nloum** [comment]
@parentelement I agree that a full release isn't necessary for it to become open source. In fact I think it would be great if you just opened sourced what you currently have. I would love to contribute rather than having to start from scratch!

**Comment by cuikp** [comment]
@parentelement @nloum 
I second the motion for a partial open source release as there's no way as yet to test how it works, bugs and all.  

In the meantime I've been working on a different kind of RTB which bypasses a lot of the difficult work by deriving from SelectableTextBlock and stacking paragraphs as a collection of them.  
(https://github.com/cuikp/AvRichTextBox) 
(https://www.nuget.org/packages/Simplecto.Avalonia.RichTextBox/)

Also very much in development still, but gets the job done for many purposes.

**Comment by parentelement** [comment]
An update for anyone who have been following.  It's a slow process because it's just me working on it but I may have at least a new update video soon as I've started actively developing again.  Just a quick note on the new things I've implemented and are working:
* Basic Import and Export for HTML, Docx, and RTF
* Bug fixes and performance improvements
* Testing and Linux
* Accessibility and Screen Reader support (Only tested with one third party screen reader software for so far).

**Comment by parentelement** [comment]
Pardon my lack of video composition skills - and my desktop is set to HDR which OBS did not translate well when recording, so colors are a little washed out and the text looks aliased, but it is crisp.  Here's a link to a quick update video I put together, and an image to show what it looks like without crummy video artifacting:
[https://www.youtube.com/watch?v=vQ6PpAO0MV0](https://www.youtube.com/watch?v=vQ6PpAO0MV0)

<img width="1401" height="1516" alt="Image" src="https://github.com/user-attachments/assets/fcb6a528-93b4-4daf-851c-855edb0f9d19" />

**Comment by parentelement** [comment]
Alright, I'd love some initial feedback.  I'm working on buttoning things up for a v0.1 beta release.  I've got some rough initial docs here.  They're not polished, just meant to show usage and capabilities:
[https://rtdocspoc.parentelement.com/](https://rtdocspoc.parentelement.com/)

**Comment by parentelement** [comment]
After what feels like days of futzing around to get an Avalonia WASM app running inside of a web application, I got a quick demo up and working.  It works in Firefox and Chrome, but not Edge (it runs from localhost on Edge but not remotely - so likely something sandbox related).  A couple of other notes:
* Styling is rough - threw it together quickly so it may show up light or dark depending on your system theme
* It takes a bit to load so be patient, you'll see the Avalonia logo while it loads the app in the background
* There is only one font available by default and I didn't load more, so the included font does not have Bold or Italic variants.  You'll see the buttons light up, but you won't see the text change.  That's the font, not the editor.
* In Firefox - using Export and choosing a format just downloads a file called "unknown".  Adding the appropriate extension to it will allow you to open and see it - This is a WASM implementation limitation that doesn't give a file save dialog.
* Justified paragraphs get "text-align: right" when exporting to HTML instead of "justify" - this is a bug.
* HTML Import freezes the editor in both Chrome and Firefox so don't try it. :)
* Lastly - the demo site/page is not responsive so not meant to be loaded on Mobile.  The editor hasn't yet been updated to accept mobile input (that's next).

I haven't dug around too much to see if there are other wasm related bugs but the quick spot checking, including adding images and import/export worked as expected:

[Avalonia Rich Text WASM Demo](https://richtextdemo.parentelement.com/)

**Comment by cuikp** [comment]
Played around a bit with the Demo.  Quite impressive as a from-scratch editor.  

As feedback, one thing I found missing that is essential for some purposes: CJK (Unicode) display and IME support.  Perhaps it actually supports Unicode but just isn't displaying because of the single Noto Mono font?  

With Undo, I notice that it seems to only honor text changes, but not font attributes (for example, after underlining, Ctrl-Z won't de-underline the text, it will just delete the last typed text, formatting and all.)

**Comment by parentelement** [comment]
> Played around a bit with the Demo. Quite impressive as a from-scratch editor.
> 
> As feedback, one thing I found missing that is essential for some purposes: CJK (Unicode) display and IME support. Perhaps it actually supports Unicode but just isn't displaying because of the single Noto Mono font?
> 
> With Undo, I notice that it seems to only honor text changes, but not font attributes (for example, after underlining, Ctrl-Z won't de-underline the text, it will just delete the last typed text, formatting and all.)

That's for the great feedback.  It does support both Unicode and RTL rendering, but it hasn't been heavily tested yet.  You won't see it in the demo as you said because of that font.  The underlying document engine does support loading your own fonts as well as Google fonts, but I haven't integrated it with the Avalonia bridge yet.  The undo system also needs an overhaul for sure.  Not all actions are captured and due to the nature of how imported data is inserted, it requires multiple undo steps to undo an import.  Copy / Paste also needs beefed up as it only copies text, and pasting doesn't support rich formatting yet, and only checks for either binary image data or text.

The editor itself is a combination of components wrapped in a WordProcessor component.  The text editing area can also standalone so it's not tied to the outer UI.

**Comment by timunie** [comment]
The feedback loops for a 3rd party control should probably better go into the repro itself. This also helps the repro owner to track.

Thank you and happy coding,
Tim

**Comment by mateli** [comment]
There is noting to watch as there is no link to the source code.
Of course you are the only one working on it when you do not give anyone else access.
I'm guessing that you will release it without commit history so anyone reviewing the code must compare a blob of changes with the upstream project?

**Comment by parentelement** [comment]
> There is noting to watch as there is no link to the source code. Of course you are the only one working on it when you do not give anyone else access. I'm guessing that you will release it without commit history so anyone reviewing the code must compare a blob of changes with the upstream project?

Sort of harsh don't you think?  All of the RichTextKit modifications are already public, with commit history:
[https://github.com/parentelement/RichTextKit](https://github.com/parentelement/RichTextKit)

There are other projects that sit on top of it, a custom controller for command processing, import/export, and of course the Avalonia integration.  Right now, for ease of development, those other things are all in a singular private repository but once I finish solidifying they'll be split out and published as NuGet packages.  The abstractions and Avalonia bridge will be open sourced.  The only piece I'm unsure of yet (and it still needs a lot of work on the theming and unification side) is the actual UI for the wordprocessor and toolbar.  So all of the capabilities will be there, and I'll likely either make the UI component itself closed / commercial, or, more likely, I'll dual license and the UI will be free for non-commercial with a separate commercial license.  However, the UI won't be necessary to get the benefit of the backend.

**Comment by parentelement** [comment]
I'm working on pulling things apart from my PoC, adding documentation, and cleaning up.  Here is the RichText.Core code that sites on top of RichTextKit.  I've published my RTK fork as a v 1.0.0 NuGet package (mistakenly as I forgot to get the versioning updated to version from the tag).  This project references the NuGet.  Just a note that this package does not have any of the Avalonia abstractions in it.  It is strictly the presentation framework on top of RTK that enables a boatload of new functionality.  Everything from the demo'd word processor and then some (for example some bug fixes, table support with cell spanning, etc) can be accomplished with this project but it's API only and draws to the Skia canvas you provide it.  There is still a bunch of cleanup to do, some TODO's to address, APIs to clean up and make more clear, functionality to add, and I need to abstract the RTK TextDocument to an interface to improve testability of the Core package.  But it is in a usable state if anyone wants to try it out.  The Wiki pages have simple documentation of all public members, classes, and interfaces.

I do not yet have public PRs enabled, but will do so when I feel like the API contracts are in a solid state, but feel free to spin up discussions if there are suggestions, or issues if you find any (I know there are a bunch of Undo fixes and enhancements I need to add to RTK).  I'll continue to work on the Avalonia abstractions for things like keymapping and once this is tidied up I'll get it configured to publish a Nuget package as well.

[https://github.com/ParentElement-LLC/RichText.Core](https://github.com/ParentElement-LLC/RichText.Core)

**Comment by parentelement** [comment]
Still a lot of restructuring to do and contracts that need to be cleaned up and changed, especially around import/export and clipboard handling but the PreRelease Nuget packages are published and as promised here is the Avalonia integration package that handles Clipboard actions, key mapping, document rendering, etc. as long as you have a SkiaCanvas to render to.  Still lots more incoming that I'm working on.  Right now RTF support is pretty limited, but I'm working on bolstering that so I can handle pasting rich content as well.  But everything you need to start testing it out with adding RichText to your Avalonia project (minus the actual UI Controls) is here.  Usage documentation for both this and the Core package are in their respective GitHub Wikis.

Right now everything is pinned to a specific Avalonia version that I've been using for testing, but as I get cleaned up and closer to a non-pre-release package I'll relax that.

[https://github.com/ParentElement-LLC/RichText.Avalonia](https://github.com/ParentElement-LLC/RichText.Avalonia)

**Comment by timunie** [comment]
@parentelement please stop using our issue tracker for your own development. This isn't the right place for this conversation. 

Thank you.

**Comment by parentelement** [comment]
> [@parentelement](https://github.com/parentelement) please stop using our issue tracker for your own development. This isn't the right place for this conversation.
> 
> Thank you.

My apologies, I picked up an old conversation and didn't realize it was an issue tracker thread and not a discussion thread.

**Comment by tomhornyak** [comment]
Is there any plans to release a WPF style rich text box? If so where is it on your road map?

**Comment by maxkatz6** [comment]
@tomhornyak yes, RichTextEditor is planned to be released soon as part of Avalonia Pro

**Comment by luthfiampas** [comment]
> [@tomhornyak](https://github.com/tomhornyak) yes, RichTextEditor is planned to be released soon as part of Avalonia Pro

I was wondering if it will support UI Virtualization out of the box?

**Comment by Gillibald** [comment]
It is fully virtualized yes
